### PR TITLE
Remove sync deadlock, unify tangle approvers and messages maps

### PR DESCRIPTION
**Pull request by zesterer**
# Description of change

Fixed 'the big' node deadlock and refactored the tangle to require fewer databases accesses and remove race conditions.

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes issue #`.

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Running the node for 15 hours without a deadlock.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-tangle/src/vertex.rs [hunk 1]
```diff
@@ -12,35 +12,63 @@ pub struct Vertex<T>
...
 
-    pub fn metadata(&self) -> &T {
-        &self.metadata
+    pub fn children_exhaustive(&self) -> bool {
+        self.children.1
     }
 
-    pub fn metadata_mut(&mut self) -> &mut T {
-        &mut self.metadata
+    /// Set the exhaustive flag. This should not be done if the vertex's children are not exhaustive.
```
Comment (zesterer) [inline bee-tangle/src/vertex.rs]: ('not' has been removed locally, btw)

File: bee-protocol/src/workers/propagator.rs [hunk 2]
```diff
@@ -52,7 +52,7 @@ async fn propagate<B: StorageBackend>(
             }
 
             // Note: There are two types of solidity carriers:
-            // * solid messages (with available history)
+            // * solid messages (with available history)(ab2
```
Comment (thibault-martinez) [inline bee-protocol/src/workers/propagator.rs]: What ?

File: bee-tangle/src/tangle.rs [hunk 3]
```diff
@@ -140,64 +135,73 @@ where
...
+                children.add_child(message_id);
+
+                // Insert cache queue entry to track eviction priority
+                cache_queue.put(parent, ());
             }
+
+            // Insert cache queue entry to track eviction priority
+            cache_queue.put(message_id, ());
+
+            drop(vertices);
```
Comment (Alex6323) [inline bee-tangle/src/tangle.rs]: seems like we can move `drop(vertices);` right before line 169

File: bee-message/src/input/mod.rs [hunk 4]
```diff
@@ -18,6 +18,7 @@ use bee_common::packable::{Packable, Read, Write};
     derive(serde::Serialize, serde::Deserialize),
     serde(tag = "type", content = "data")
 )]
+#[allow(clippy::upper_case_acronyms)]
```
Comment (thibault-martinez) [inline bee-message/src/input/mod.rs]: I renamed the enum variant, you can rebase and remove the allow

File: bee-tangle/src/tangle.rs (hunk 3 above)
Comment (zesterer) [reply bee-tangle/src/tangle.rs]: We can now, yes

File: bee-protocol/src/workers/propagator.rs (hunk 2 above)
Comment (zesterer) [reply bee-protocol/src/workers/propagator.rs]: My keyboard has a habit of occasionally spewing random letters into things. I'm not sure why, it's really annoying.

**Review by Alex6323 — APPROVED** [review]

**Pull request merged**